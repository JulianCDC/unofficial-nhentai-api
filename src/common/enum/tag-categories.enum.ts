export enum CATEGORIES {
  tags,
  artists,
  characters,
  parodies,
  groups,
}
