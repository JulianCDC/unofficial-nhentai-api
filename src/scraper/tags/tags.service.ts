import { Injectable, Logger } from '@nestjs/common';
import { InjectModel } from '@nestjs/mongoose';
import { Model } from 'mongoose';
import { Tag } from '../../common/schemas/tag.schema';
import { Tag as TagEntity } from '../../common/types/tag.entity';
import { Cron, CronExpression } from '@nestjs/schedule';
import { getPages } from '../../common/helpers/get-pagination.helper';
import { HtmlParserService } from '../../html-parser/html-parser/html-parser.service';
import { parseTags } from '../../common/helpers/parse-tags.helper';

@Injectable()
export class TagsService {
  private CATEGORIES = ['tags', 'artists', 'characters', 'parodies', 'groups'];
  private readonly logger = new Logger(TagsService.name);

  constructor(
    @InjectModel(Tag.name) private tagModel: Model<Tag>,
    private htmlParser: HtmlParserService,
  ) {}

  @Cron(CronExpression.EVERY_DAY_AT_MIDNIGHT)
  parseAllTags(): void {
    this.CATEGORIES.forEach(async category => {
      const tags: TagEntity[] = [];
      const $ = await this.htmlParser.parse(`https://nhentai.net/${category}`);

      const numberOfPages = getPages($, 0);

      for (let i = 1; i <= numberOfPages; i++) {
        const uri = `https://nhentai.net/${category}/?page=${i}`;
        const $page = await this.htmlParser.parse(uri);

        tags.push(...parseTags($page('.tag'), $page));
      }

      tags.forEach(tag => {
        this.tagModel.findOneAndUpdate(
          { name: tag.name },
          { ...tag, category: category },
          {
            upsert: true,
          },
          err => {
            this.logger.error(err);
          },
        );
      });
    });
  }
}
