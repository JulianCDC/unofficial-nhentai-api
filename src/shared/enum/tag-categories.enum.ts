export enum CATEGORIES {
  tags = 'tags',
  artists = 'artists',
  characters = 'characters',
  parodies = 'parodies',
  groups = 'groups',
}
