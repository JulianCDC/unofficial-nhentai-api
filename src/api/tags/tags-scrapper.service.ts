import { Injectable, Logger } from '@nestjs/common';
import { InjectModel } from '@nestjs/mongoose';
import { Cron, CronExpression } from '@nestjs/schedule';
import { Model } from 'mongoose';
import { CATEGORIES } from 'src/shared/enum/tag-categories.enum';
import { HtmlTag } from 'src/shared/types/html-tag.entity';
import { HtmlParserService } from '../../html-parser/html-parser/html-parser.service';
import { getPages } from '../../shared/helpers/get-pagination.helper';
import { parseTags } from '../../shared/helpers/parse-tags.helper';
import { TagDocument } from '../../shared/schemas/tag.schema';

@Injectable()
export class TagsScrapperService {
  private readonly categories = Object.values(CATEGORIES);
  private readonly logger = new Logger(TagsScrapperService.name);

  constructor(
    @InjectModel(TagDocument.name) private tagModel: Model<TagDocument>,
    private htmlParser: HtmlParserService,
  ) {}

  @Cron(CronExpression.EVERY_WEEK)
  parseAllTags(): void {
    this.categories.forEach(async category => {
      const tags: HtmlTag[] = [];
      const $ = await this.htmlParser.parse(`https://nhentai.net/${category}`);

      const numberOfPages = getPages($, 0);

      for (let i = 1; i <= numberOfPages; i++) {
        const uri = `https://nhentai.net/${category}/?page=${i}`;
        const $page = await this.htmlParser.parse(uri);

        tags.push(...parseTags($page('.tag'), $page));
      }

      tags.forEach(tag => {
        this.tagModel.findOneAndUpdate(
          { name: tag.name },
          { ...tag, category: category },
          {
            upsert: true,
          },
          err => {
            this.logger.error(err);
          },
        );
      });
    });
  }
}
