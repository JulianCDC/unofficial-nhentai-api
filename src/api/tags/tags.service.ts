import { Injectable, NotFoundException } from '@nestjs/common';
import { InjectModel } from '@nestjs/mongoose';
import { Model } from 'mongoose';
import { Tag } from 'src/shared/types/tag.entity';
import { CATEGORIES } from '../../shared/enum/tag-categories.enum';
import { TagDocument } from '../../shared/schemas/tag.schema';

@Injectable()
export class TagsService {
  static PAGE_SIZE = 120;

  constructor(
    @InjectModel(TagDocument.name) private tagModel: Model<TagDocument>,
  ) {}

  async getTagById(id: number): Promise<Tag> {
    const tag = await this.tagModel
      .findOne({ id })
      .select('-__v -_id')
      .exec();

    if (!tag) {
      throw new NotFoundException();
    }

    return tag;
  }

  // TODO: add pagination
  // TODO: sort before slicing
  async search(searchQuery: string, category?: string): Promise<Tag[]> {
    const nameRegexp = new RegExp(`^.*${searchQuery}.*$`);

    let tags: Tag[];

    if (category) {
      tags = await this.tagModel
        .find({ name: nameRegexp, category: category })
        .select('-__v -_id')
        .limit(10);
    } else {
      tags = await this.tagModel
        .find({ name: nameRegexp })
        .select('-__v -_id')
        .limit(10);
    }

    return tags.sort((a, b) => {
      const aName = a.name;
      const bName = b.name;

      return aName.indexOf(searchQuery) - bName.indexOf(searchQuery);
    });
  }

  async getTagsByPopularity(
    category: CATEGORIES,
    page: number,
  ): Promise<Tag[]> {
    return this.tagModel
      .find({ category: category.toString() })
      .select('-__v -_id')
      .sort({ tagged: 'desc' })
      .skip((page - 1) * TagsService.PAGE_SIZE)
      .limit(120)
      .exec();
  }

  async getTags(category: CATEGORIES, page: number): Promise<Tag[]> {
    return this.tagModel
      .find({ category: category.toString() })
      .select('-__v -_id')
      .sort({ name: 'asc' })
      .skip((page - 1) * TagsService.PAGE_SIZE)
      .limit(120)
      .exec();
  }

  async getPageCount(category: CATEGORIES): Promise<number> {
    const modelCount = await this.tagModel
      .countDocuments({ category: category.toString() })
      .exec();

    return Math.ceil(modelCount / TagsService.PAGE_SIZE);
  }

  async getTagsByLetter(category: CATEGORIES, letter: string): Promise<Tag[]> {
    const nameRegexp = new RegExp(`^${letter}`, 'i');

    return this.tagModel
      .find({
        category: category.toString(),
        name: nameRegexp,
      })
      .select('-__v -_id')
      .sort({ name: 'asc' })
      .exec();
  }
}
