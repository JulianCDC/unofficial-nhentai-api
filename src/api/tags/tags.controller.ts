import {
  CacheInterceptor,
  Controller,
  Get,
  NotFoundException,
  Param,
  Query,
  UseInterceptors,
} from '@nestjs/common';
import {
  ApiBadRequestResponse,
  ApiNotFoundResponse,
  ApiOperation,
  ApiTags,
} from '@nestjs/swagger';
import { Tag } from 'src/shared/types/tag.entity';
import { CATEGORIES } from '../../shared/enum/tag-categories.enum';
import { TagsService } from './tags.service';
import { GetCategoryPageByLetterDto } from './types/get-category-page-by-letter.dto';
import { GetCategoryPageDto } from './types/get-category-page.dto';
import { GetCategoryDto } from './types/get-category.dto';
import { GetTagByIdDto } from './types/get-tag-by-id.dto';
import { SearchCategoryDto } from './types/search-category.dto';
import { TagListEntity } from './types/tag-list.entity';

@ApiTags('tags')
@Controller('tags')
@UseInterceptors(CacheInterceptor)
export class TagsController {
  constructor(private tagsService: TagsService) {}

  @ApiOperation({
    summary: 'get tag infos for requested id',
  })
  @ApiBadRequestResponse()
  @ApiNotFoundResponse({ description: 'no tag found with this id' })
  @Get(':id/details')
  async getTagById(@Param() params: GetTagByIdDto): Promise<Tag> {
    return this.tagsService.getTagById(params.id);
  }

  @ApiOperation({
    summary: 'search for a tag by name and optionaly by category',
    description:
      'returns a list sorted depending on the position of the search query in the tag name',
  })
  @ApiBadRequestResponse()
  @Get('search')
  async searchTag(@Query() query: SearchCategoryDto): Promise<Tag[]> {
    return this.tagsService.search(query.q, query.page, query.category);
  }

  @ApiOperation({
    summary: 'get possible values for category',
  })
  @Get('categories')
  getCategories(): string[] {
    return Object.values(CATEGORIES);
  }

  @ApiOperation({
    summary: 'get all existing tags in specified category',
  })
  @ApiBadRequestResponse()
  @ApiNotFoundResponse({ description: 'requested page does not exist' })
  @Get('list/:category')
  async getTagsByCategory(
    @Param() params: GetCategoryDto,
    @Query() query: GetCategoryPageDto,
  ): Promise<TagListEntity> {
    let tags: Tag[];
    const numberOfPages = await this.tagsService.getPageCount(params.category);

    if (query.page > numberOfPages) {
      throw new NotFoundException();
    }

    if (query.popular) {
      tags = await this.tagsService.getTagsByPopularity(
        params.category,
        query.page,
      );
    } else {
      tags = await this.tagsService.getTags(params.category, query.page);
    }

    return {
      data: tags,
      pagination: {
        page: query.page,
        total: numberOfPages,
      },
    };
  }

  @ApiOperation({
    summary:
      'get all existing tags starting by specified letter in specified category',
  })
  @ApiBadRequestResponse()
  @Get('list/:category/:letter')
  async getTagsByCategoryAndLetter(
    @Param() params: GetCategoryPageByLetterDto,
  ): Promise<Tag[]> {
    return this.tagsService.getTagsByLetter(params.category, params.letter);
  }
}
