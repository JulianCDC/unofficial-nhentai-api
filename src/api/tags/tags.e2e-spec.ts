jest.mock('./tags.service');

import { INestApplication } from '@nestjs/common';
import { getModelToken } from '@nestjs/mongoose';
import { Test } from '@nestjs/testing';
import { random } from 'faker';
import { curry } from 'lodash';
import * as request from 'supertest';
import { createTestingApp, testRequests } from '../../../test/helpers/e2e';
import { CATEGORIES } from '../../shared/enum/tag-categories.enum';
import { TagDocument } from '../../shared/schemas/tag.schema';
import { TagsModule } from './tags.module';
import { TagsService } from './tags.service';
import {
  mockedTags,
  TagsService as MockedTagsService,
} from './__mocks__/tags.service';

describe('TagsController (e2e)', () => {
  let app: INestApplication;
  let tagsService: TagsService;

  beforeAll(async () => {
    const moduleRef = await Test.createTestingModule({
      imports: [TagsModule],
    })
      .overrideProvider(TagsService)
      .useClass(MockedTagsService)
      .overrideProvider(getModelToken(TagDocument.name))
      .useValue({})
      .compile();

    tagsService = moduleRef.get(TagsService);

    app = createTestingApp(moduleRef);
    await app.init();
  });

  describe('/tags/list/:category (GET)', () => {
    it('sort by popularity', async () => {
      const result = await tagsService.getTagsByPopularity(
        CATEGORIES.artists,
        1,
      );

      return request(app.getHttpServer())
        .get('/tags/list/artists?popular=true&page=1')
        .expect(200)
        .expect({
          data: result,
          pagination: {
            page: 1,
            total: await tagsService.getPageCount(CATEGORIES.artists),
          },
        });
    });

    it('sort by name', async () => {
      const result = await tagsService.getTags(CATEGORIES.artists, 1);

      return request(app.getHttpServer())
        .get('/tags/list/artists?popular=false&page=1')
        .expect(200)
        .expect({
          data: result,
          pagination: {
            page: 1,
            total: await tagsService.getPageCount(CATEGORIES.artists),
          },
        });
    });

    it('throws 404 error when requested page does not exist', async () => {
      const numberOfPages = await tagsService.getPageCount(CATEGORIES.artists);

      return request(app.getHttpServer())
        .get(`/tags/list/artists?popular=false&page=${numberOfPages + 1}`)
        .expect(404);
    });

    it('throws 400 error on wrong parameters values', async () => {
      const httpServer = app.getHttpServer();
      const badRequests: string[] = [
        '/tags/list/notacategory?popular=false&page=1',
        '/tags/list/artists?popular=false&page=0',
        '/tags/list/artists?popular=false&page=1.50',
        '/tags/list/artists?popular=false&page=notanumber',
        '/tags/list/artists?popular=false&page=-1',
      ];

      await testRequests(httpServer, badRequests, 400);
    });

    it('converts parameters types', async () => {
      const httpServer = app.getHttpServer();
      const byPopularity = curry(tagsService.getTagsByPopularity)(
        CATEGORIES.artists,
      );
      const byName = curry(tagsService.getTags)(CATEGORIES.artists);

      const shouldOrderByPopularity: string[] = [
        '/tags/list/artists?popular=true&page=1',
        '/tags/list/artists?popular=1&page=1',
        '/tags/list/artists?popular=anything&page=1',
      ];
      const shouldOrderByName: string[] = [
        '/tags/list/artists?popular=false&page=1',
        '/tags/list/artists?page=1',
      ];
      const shouldWorks: string[] = [
        '/tags/list/artists',
        '/tags/list/artists?page=1',
      ];

      for (const req of shouldOrderByPopularity) {
        await request(httpServer)
          .get(req)
          .expect(200)
          .then(async res => {
            expect(res.body.data).toStrictEqual(await byPopularity(1));
          });
      }

      for (const req of shouldOrderByName) {
        await request(httpServer)
          .get(req)
          .expect(200)
          .then(async res => {
            expect(res.body.data).toStrictEqual(await byName(1));
          });
      }

      for (const req of shouldWorks) {
        await request(httpServer)
          .get(req)
          .expect(200);
      }
    });
  });

  describe('/tags/list/:category/:letter (GET)', () => {
    it('returns all tags starting with specified letter, sorted by name', async () => {
      const result = await tagsService.getTagsByLetter(CATEGORIES.artists, 'A');

      return request(app.getHttpServer())
        .get('/tags/list/artists/a')
        .expect(200)
        .expect(result);
    });

    it('is case insensitive', async () => {
      const result = await tagsService.getTagsByLetter(CATEGORIES.artists, 'A');

      await request(app.getHttpServer())
        .get('/tags/list/artists/a')
        .expect(200)
        .expect(result);

      await request(app.getHttpServer())
        .get('/tags/list/artists/A')
        .expect(200)
        .expect(result);
    });

    it('throws 400 error on wrong parameters values', async () => {
      const httpServer = app.getHttpServer();
      const badRequests: string[] = [
        '/tags/list/notacategory/a',
        '/tags/list/artists/notaletter',
      ];

      await testRequests(httpServer, badRequests, 400);
    });
  });

  describe('/tags/details/:id (GET)', () => {
    it('returns requested tag', async () => {
      const result = await tagsService.getTagById(
        random.arrayElement(mockedTags).id,
      );

      await request(app.getHttpServer())
        .get(`/tags/details/${result.id}`)
        .expect(200)
        .expect(result);
    });

    it('returns 404 error when tag does not exist', async () => {
      await request(app.getHttpServer())
        .get('/tags/details/999999')
        .expect(404);
    });

    it('validate and convert parameters', async () => {
      const httpServer = app.getHttpServer();
      const badRequests = [
        '/tags/details/0',
        '/tags/details/bite',
        '/tags/details/NaN',
      ];
      const goodRequests = [
        `/tags/details/${random.arrayElement(mockedTags).id}`,
      ];

      await testRequests(httpServer, badRequests, 400);
      await testRequests(httpServer, goodRequests, 200);
    });
  });

  describe('/tags/search', () => {
    it.todo('returns tags including search query');

    it.todo('returns tags including search query and category');

    it.todo('returns empty array if nothing found');

    it.todo('validate and convert parameters');
  });

  afterAll(async () => {
    await app.close();
  });
});
