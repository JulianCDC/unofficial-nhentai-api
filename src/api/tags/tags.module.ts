import { CacheModule, Module } from '@nestjs/common';
import { MongooseModule } from '@nestjs/mongoose';
import { HtmlParserModule } from 'src/html-parser/html-parser.module';
import { TagDocument, TagSchema } from '../../shared/schemas/tag.schema';
import { TagsScrapperService } from './tags-scrapper.service';
import { TagsController } from './tags.controller';
import { TagsService } from './tags.service';

@Module({
  imports: [
    MongooseModule.forFeature([{ name: TagDocument.name, schema: TagSchema }]),
    CacheModule.register(),
    HtmlParserModule,
  ],
  controllers: [TagsController],
  providers: [TagsService, TagsScrapperService],
})
export class TagsModule {}
