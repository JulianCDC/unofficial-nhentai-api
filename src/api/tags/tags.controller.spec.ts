import { CacheModule, NotFoundException } from '@nestjs/common';
import { getModelToken } from '@nestjs/mongoose';
import { Test } from '@nestjs/testing';
import { random } from 'faker';
import { chunk } from 'lodash';
import { Tag } from 'src/shared/types/tag.entity';
import { generateRandomTags } from '../../../test/mocks/tags.mock';
import { CATEGORIES } from '../../shared/enum/tag-categories.enum';
import { TagDocument } from '../../shared/schemas/tag.schema';
import { TagsController } from './tags.controller';
import { TagsService } from './tags.service';
import { TagListEntity } from './types/tag-list.entity';

describe('CategoriesController', () => {
  let tagsController: TagsController;
  let tagsService: TagsService;

  let mockTags: Tag[] = [];
  const numberOfPages = async (category: CATEGORIES) => {
    const filteredTags = mockTags.filter(
      tag => tag.category === category.toString(),
    );

    return Math.ceil(filteredTags.length / pageSize);
  };

  const pageSize = 120;

  beforeEach(async () => {
    const moduleRef = await Test.createTestingModule({
      controllers: [TagsController],
      providers: [
        TagsService,
        {
          provide: getModelToken(TagDocument.name),
          useValue: {},
        },
      ],
      imports: [CacheModule.register({ ttl: 1 })],
    }).compile();

    tagsService = moduleRef.get(TagsService);
    tagsController = moduleRef.get(TagsController);

    mockTags = generateRandomTags(1000);
  });

  describe('getCategories', () => {
    it('should returns available categories', () => {
      const results = ['tags', 'artists', 'characters', 'parodies', 'groups'];

      expect(tagsController.getCategories()).toStrictEqual(results);
    });
  });

  describe('getTagsByCategory', () => {
    const getTags = async (category: CATEGORIES, page: number) => {
      return mockTags
        .filter(data => data.category === category)
        .sort((a, b) => {
          if (a.name > b.name) return 1;
          if (a.name < b.name) return -1;

          return 0;
        })
        .slice(page, page + pageSize);
    };

    const getTagsByPopularity = async (category: CATEGORIES, page: number) => {
      return mockTags
        .filter(data => data.category === category.toString())
        .sort((a, b) => a.tagged - b.tagged)
        .slice(page, page + pageSize);
    };

    beforeEach(() => {
      jest.spyOn(tagsService, 'getPageCount').mockImplementation(numberOfPages);

      jest.spyOn(tagsService, 'getTags').mockImplementation(getTags);

      jest
        .spyOn(tagsService, 'getTagsByPopularity')
        .mockImplementation(getTagsByPopularity);
    });

    afterEach(() => {
      jest.restoreAllMocks();
    });

    it('get values by name in first page', async () => {
      const results: TagListEntity = {
        data: await getTags(CATEGORIES.artists, 1),
        pagination: {
          page: 1,
          total: await numberOfPages(CATEGORIES.artists),
        },
      };

      expect(
        await tagsController.getTagsByCategory(
          { category: CATEGORIES.artists },
          { page: 1 },
        ),
      ).toEqual(results);
    });

    it('get value by popularity in first page', async () => {
      const results: TagListEntity = {
        data: await getTagsByPopularity(CATEGORIES.artists, 1),
        pagination: {
          page: 1,
          total: await numberOfPages(CATEGORIES.artists),
        },
      };

      expect(
        await tagsController.getTagsByCategory(
          { category: CATEGORIES.artists },
          { page: 1, popular: true },
        ),
      ).toEqual(results);
    });

    it.each(chunk(Object.keys(CATEGORIES), 1))(
      `correctly filters values for all categories: %s`,
      async category => {
        const currentCategory = CATEGORIES[category];

        const tags = await tagsController
          .getTagsByCategory(
            {
              category: currentCategory,
            },
            {
              page: 1,
            },
          )
          .then(categories => categories.data);

        tags.forEach(tag => {
          expect(tag.category).toBe(category);
        });
      },
    );

    it('get values by name in random page', async () => {
      const numberOfArtistPage = await numberOfPages(CATEGORIES.artists);

      const randomPage = random.number(numberOfArtistPage);
      const results: TagListEntity = {
        data: await getTags(CATEGORIES.artists, randomPage),
        pagination: {
          page: randomPage,
          total: numberOfArtistPage,
        },
      };

      expect(
        await tagsController.getTagsByCategory(
          { category: CATEGORIES.artists },
          { page: randomPage },
        ),
      ).toEqual(results);
    });

    it('throws an error when requested page is above total number of pages', async () => {
      await expect(
        tagsController.getTagsByCategory(
          { category: CATEGORIES.artists },
          { page: await numberOfPages(CATEGORIES.artists) },
        ),
      ).resolves.toBeDefined();

      await expect(
        tagsController.getTagsByCategory(
          { category: CATEGORIES.artists },
          { page: (await numberOfPages(CATEGORIES.artists)) + 1 },
        ),
      ).rejects.toThrowError(NotFoundException);
    });
  });

  describe('getTagsByCategoryAndLetter', () => {
    const getTagsByLetter = async (category: CATEGORIES, letter: string) => {
      return mockTags
        .filter(
          data =>
            data.category === category &&
            data.name.toLowerCase().startsWith(letter.toLowerCase()),
        )
        .sort((a, b) => {
          if (a.name > b.name) return 1;
          if (a.name < b.name) return -1;

          return 0;
        });
    };

    beforeEach(() => {
      mockTags = generateRandomTags(1000);

      jest
        .spyOn(tagsService, 'getTagsByLetter')
        .mockImplementation(getTagsByLetter);
    });

    afterEach(() => {
      jest.restoreAllMocks();
    });

    it('returns all values starting with letter', async () => {
      const results = await getTagsByLetter(CATEGORIES.artists, 'A');

      await expect(
        tagsController.getTagsByCategoryAndLetter({
          category: CATEGORIES.artists,
          letter: 'A',
        }),
      ).resolves.toEqual(results);
    });

    it('returns empty array when letter does not exists', async () => {
      await expect(
        tagsController.getTagsByCategoryAndLetter({
          category: CATEGORIES.artists,
          letter: 'not a letter',
        }),
      ).resolves.toEqual([]);
    });
  });

  describe('getTagById', () => {
    let tag: Tag;

    beforeEach(() => {
      tag = random.arrayElement(mockTags);

      jest.spyOn(tagsService, 'getTagById').mockImplementation(async () => tag);
    });

    it('returns tag details', async () => {
      await expect(
        tagsController.getTagById({ id: tag.id }),
      ).resolves.toStrictEqual(tag);
    });
  });

  describe('searchTag', () => {
    const filterTags = async (query: string, category: CATEGORIES) => {
      const matchRegexp = new RegExp(query, 'gi');

      return mockTags.filter(
        tag =>
          tag.category === category.toString() && tag.name.match(matchRegexp),
      );
    };

    beforeEach(() => {
      jest.spyOn(tagsService, 'search').mockImplementation(filterTags);
    });

    it('returns tags matching query', async () => {
      await expect(
        tagsController.searchTag({ q: 'ab', category: CATEGORIES.artists }),
      ).resolves.toStrictEqual(await filterTags('ab', CATEGORIES.artists));
    });
  });
});
