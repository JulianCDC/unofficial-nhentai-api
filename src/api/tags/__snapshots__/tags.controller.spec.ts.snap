// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CategoriesController getTagById returns tag details 1`] = `
Object {
  "category": "parodies",
  "id": 225709,
  "name": "itaque",
  "tagged": 4455,
  "uri": "https://berenice.net",
}
`;

exports[`CategoriesController getTagsByCategory correctly filters values for all categories: artists 1`] = `
Object {
  "data": Array [
    Object {
      "category": "artists",
      "id": 128124,
      "name": "esse",
      "tagged": 99718,
      "uri": "https://torey.com",
    },
    Object {
      "category": "artists",
      "id": 39054,
      "name": "et",
      "tagged": 89460,
      "uri": "https://arvel.net",
    },
    Object {
      "category": "artists",
      "id": 878116,
      "name": "ipsam",
      "tagged": 68521,
      "uri": "https://connor.info",
    },
    Object {
      "category": "artists",
      "id": 186260,
      "name": "repellat",
      "tagged": 14675,
      "uri": "http://audrey.info",
    },
    Object {
      "category": "artists",
      "id": 691876,
      "name": "tempore",
      "tagged": 95788,
      "uri": "http://kaci.biz",
    },
  ],
  "pagination": Object {
    "page": 1,
    "total": 21,
  },
}
`;

exports[`CategoriesController getTagsByCategory correctly filters values for all categories: characters 1`] = `
Object {
  "data": Array [
    Object {
      "category": "characters",
      "id": 457204,
      "name": "accusantium",
      "tagged": 53441,
      "uri": "https://tatyana.info",
    },
    Object {
      "category": "characters",
      "id": 593064,
      "name": "ipsum",
      "tagged": 5924,
      "uri": "https://marco.com",
    },
    Object {
      "category": "characters",
      "id": 846310,
      "name": "ut",
      "tagged": 66974,
      "uri": "http://tre.name",
    },
  ],
  "pagination": Object {
    "page": 1,
    "total": 23,
  },
}
`;

exports[`CategoriesController getTagsByCategory correctly filters values for all categories: groups 1`] = `
Object {
  "data": Array [
    Object {
      "category": "groups",
      "id": 518152,
      "name": "ut",
      "tagged": 80275,
      "uri": "http://augusta.net",
    },
  ],
  "pagination": Object {
    "page": 1,
    "total": 19,
  },
}
`;

exports[`CategoriesController getTagsByCategory correctly filters values for all categories: parodies 1`] = `
Object {
  "data": Array [
    Object {
      "category": "parodies",
      "id": 800743,
      "name": "et",
      "tagged": 14038,
      "uri": "https://clotilde.net",
    },
  ],
  "pagination": Object {
    "page": 1,
    "total": 20,
  },
}
`;

exports[`CategoriesController getTagsByCategory correctly filters values for all categories: tags 1`] = `
Object {
  "data": Array [],
  "pagination": Object {
    "page": 1,
    "total": 17,
  },
}
`;

exports[`CategoriesController getTagsByCategory get value by popularity in first page 1`] = `
Object {
  "data": Array [
    Object {
      "category": "artists",
      "id": 128124,
      "name": "esse",
      "tagged": 99718,
      "uri": "https://torey.com",
    },
    Object {
      "category": "artists",
      "id": 691876,
      "name": "tempore",
      "tagged": 95788,
      "uri": "http://kaci.biz",
    },
    Object {
      "category": "artists",
      "id": 39054,
      "name": "et",
      "tagged": 89460,
      "uri": "https://arvel.net",
    },
    Object {
      "category": "artists",
      "id": 878116,
      "name": "ipsam",
      "tagged": 68521,
      "uri": "https://connor.info",
    },
    Object {
      "category": "artists",
      "id": 186260,
      "name": "repellat",
      "tagged": 14675,
      "uri": "http://audrey.info",
    },
  ],
  "pagination": Object {
    "page": 1,
    "total": 21,
  },
}
`;

exports[`CategoriesController getTagsByCategory get values by name in any page 1`] = `
Object {
  "data": Array [
    Object {
      "category": "artists",
      "id": 748164,
      "name": "aperiam",
      "tagged": 75014,
      "uri": "https://yasmin.name",
    },
    Object {
      "category": "artists",
      "id": 916860,
      "name": "magni",
      "tagged": 52479,
      "uri": "http://aron.net",
    },
  ],
  "pagination": Object {
    "page": 2,
    "total": 21,
  },
}
`;

exports[`CategoriesController getTagsByCategory get values by name in first page 1`] = `
Object {
  "data": Array [
    Object {
      "category": "artists",
      "id": 128124,
      "name": "esse",
      "tagged": 99718,
      "uri": "https://torey.com",
    },
    Object {
      "category": "artists",
      "id": 39054,
      "name": "et",
      "tagged": 89460,
      "uri": "https://arvel.net",
    },
    Object {
      "category": "artists",
      "id": 878116,
      "name": "ipsam",
      "tagged": 68521,
      "uri": "https://connor.info",
    },
    Object {
      "category": "artists",
      "id": 186260,
      "name": "repellat",
      "tagged": 14675,
      "uri": "http://audrey.info",
    },
    Object {
      "category": "artists",
      "id": 691876,
      "name": "tempore",
      "tagged": 95788,
      "uri": "http://kaci.biz",
    },
  ],
  "pagination": Object {
    "page": 1,
    "total": 21,
  },
}
`;

exports[`CategoriesController getTagsByCategoryAndLetter returns all values starting with letter 1`] = `
Array [
  Object {
    "category": "artists",
    "id": 22330,
    "name": "ab",
    "tagged": 51935,
    "uri": "https://ozella.net",
  },
  Object {
    "category": "artists",
    "id": 918732,
    "name": "accusantium",
    "tagged": 54954,
    "uri": "http://tierra.net",
  },
  Object {
    "category": "artists",
    "id": 748164,
    "name": "aperiam",
    "tagged": 75014,
    "uri": "https://yasmin.name",
  },
  Object {
    "category": "artists",
    "id": 550947,
    "name": "aut",
    "tagged": 57974,
    "uri": "http://genoveva.name",
  },
  Object {
    "category": "artists",
    "id": 517593,
    "name": "aut",
    "tagged": 93487,
    "uri": "https://oswaldo.name",
  },
]
`;

exports[`CategoriesController searchTag returns tags matching query 1`] = `
Array [
  Object {
    "category": "artists",
    "id": 878116,
    "name": "ipsam",
    "tagged": 68521,
    "uri": "https://connor.info",
  },
  Object {
    "category": "artists",
    "id": 593480,
    "name": "hic",
    "tagged": 59911,
    "uri": "http://adrienne.org",
  },
  Object {
    "category": "artists",
    "id": 239847,
    "name": "quis",
    "tagged": 97001,
    "uri": "http://rosella.net",
  },
  Object {
    "category": "artists",
    "id": 233622,
    "name": "omnis",
    "tagged": 96766,
    "uri": "https://dudley.com",
  },
  Object {
    "category": "artists",
    "id": 748164,
    "name": "aperiam",
    "tagged": 75014,
    "uri": "https://yasmin.name",
  },
  Object {
    "category": "artists",
    "id": 916860,
    "name": "magni",
    "tagged": 52479,
    "uri": "http://aron.net",
  },
  Object {
    "category": "artists",
    "id": 538830,
    "name": "molestias",
    "tagged": 24621,
    "uri": "http://ryley.biz",
  },
  Object {
    "category": "artists",
    "id": 322680,
    "name": "temporibus",
    "tagged": 69805,
    "uri": "https://sam.com",
  },
  Object {
    "category": "artists",
    "id": 918732,
    "name": "accusantium",
    "tagged": 54954,
    "uri": "http://tierra.net",
  },
]
`;
